plugin/evm/statesync/dynamic_syncer.go

```diff
@@ -0,0 +1,236 @@
+// (c) 2021-2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package statesync
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/ids"
+	// xsync "github.com/ava-labs/avalanchego/x/sync"
+	"github.com/ava-labs/coreth/core/types"
+	"github.com/ava-labs/coreth/peer"
+	"github.com/ava-labs/libevm/common"
+	"github.com/ava-labs/libevm/core/rawdb"
+	"github.com/ava-labs/libevm/ethdb"
+	"github.com/ava-labs/libevm/log"
+)
+
+type SyncBlockRequest uint8
+
+const (
+	// Constants to identify block requests
+	VerifySyncBlockRequest SyncBlockRequest = iota + 1
+	AcceptSyncBlockRequest
+	RejectSyncBlockRequest
+
+	// Dynamic state switches state root occasionally
+	// Buffer must be large enough to
+	pivotInterval = 128
+	bufferSize    = 3 * pivotInterval
+
+	snapSyncType = "snap"
+)
+
+var (
+	queueOverflowError = errors.New("Snap sync queue overflow")
+
+	_ manager = &snapManager{}
+	// _ manager = &xsync.Manager{} TODO: ensure this matches
+)
+
+type queueElement struct {
+	block    *types.Block
+	req      SyncBlockRequest
+	resolver func() error
+}
+
+// Interface used by x/sync
+type manager interface {
+	// Stops syncing progress
+	Close()
+
+	// Nonnil if a fatal error occurred
+	Error() error
+
+	// Initiates state sync in the background
+	Start(ctx context.Context) error
+
+	// UpdateSyncTarget updates the sync target to the given root hash.
+	// Can be called concurrently?
+	UpdateSyncTarget(syncTargetRoot common.Hash) error
+
+	// Wait blocks until one of the following occurs:
+	// - sync is complete.
+	// - sync fatally errored.
+	// - [ctx] is canceled.
+	// If [ctx] is canceled, returns [ctx].Err().
+	Wait(ctx context.Context) error
+}
+
+type DynamicSyncConfig struct {
+	// ChainDB is the database that the downloader will use to store the synced state
+	ChainDB ethdb.Database
+	// PivotBlock is the block that the downloader will use as the pivot block
+	FirstPivotBlock *types.Block
+	// Scheme is the state scheme that the downloader will use to store the synced state
+	Scheme string
+	// SyncQueueLock is the lock that will be used to protect the block buffer
+	SyncQueueLock *sync.Mutex
```

Comment:
I kind of find this lock a bit of a bother / hack:
- First, it requires someone else to initialize & provide a lock for this struct,
- There's an asymetrical call to Unlock in Close
- Some exported methods assume the lock is held vs. some others don't.

mentioned in another comment but I think we can split a CallbackQueue out of this struct and move the logic about deciding when to empty the queue and pivot the manager to a different place.